[doc] Initial PARTITION BY documentation (#30485)
MaterializeInc/materialize: doc/user/content/transform-data/patterns/partition-by.md

---

> title: "PARTITION BY"

Author: This doc covers both the `PARTITION BY` syntax and the filter pushdown optimization, which are strictly speaking two separate things but likely to be used together I think?

Reviewer: 👍 Works for me.
We could title it PARTITION BY/filter pushdown optimization or Filter pushdown optimzation (PARTITION BY) -- not quite sure the naming scheme of the pages in the patterns section.

Reviewer: Oh, should we at least mention add a link to this page in the temporal filters pattern page?

Author: Great - took another stab at the name. (The all-uppercase version seemed out of step with the other guides.)

It does make sense to link from that other doc, but I had a hard time finding a way to do it that felt appropriate for a private preview feature. I left a TODO for now, but I'm happy for other suggestions.

Reviewer: ❤️ the name.
As for the temporal filters page, maybe add a "See also [Partitioning and filter pushdown](/transform-data/patterns/partition-by/)" at the end of the page -- if we want to get more users to turn on the feature for them.

description: "Specifying the sort order of the data."
aliases:
  - /guides/partition-by/
  - /sql/patterns/partition-by/
menu:
  main:
    parent: 'sql-patterns'
---

A few types of Materialize collections are durably written to storage: [materialized views](/sql/create-materialized-view/), [tables](/sql/create-table), and [sources](/sql/create-source).

Internally, Materialize stores these durable collections in an [LSM-tree](https://en.wikipedia.org/wiki/Log-structured_merge-tree)-like structure. Each collection is made up of a set of
**runs** of data, each run is sorted and then split up into individual **parts**, and those parts are written to object storage and retrieved only when necessary to satisfy a query. Materialize will also periodically **compact** the data it stores to consolidate small parts into larger ones or discard deleted rows.

> Materialize lets you specify the ordering it will use to sort these runs of data internally. A well-chosen sort order can unlock optimizations like [filter pushdown](#filter-pushdown), which in turn can make queries and other operations more efficient.

Author: It seems like the "patterns" section is where we put this sort of doc that covers features that aren't specific to a single type and involve some discussion of implementation details, but let me know if you'd like me to move it!

Reviewer: I too am a little confused by what goes into our patterns section, but the pattern of sort order + filter pushdown seems like it would go here.

Author: Cool, will leave it alone for now then!

> Materialize lets you specify the ordering it will use to sort these runs of data internally. A well-chosen sort order can unlock optimizations like [filter pushdown](#filter-pushdown), which in turn can make queries and other operations more efficient.

Author: I think it's important to describe the default behaviour alongside the option itself, but I've tried to rewrite a bit to provide the same information in the intro! (And I've mentioned the option here and a bunch of other places where it felt appropriate to be concrete... thanks for the callout.)

## Syntax

> The option `PARTITION BY <column list>` declares that a [materialized view](/sql/create-materialized-view/#with_options), [table](/sql/create-table/#with_options), or source table should be ordered by the listed columns. For example, a table that stores an append-only collection of events may look like:

Author: I've updated the MV docs as part of this PR as well.

I have not updated the table docs yet: those do not have a separate section for "with options" yet. I'm happy to add such a section and move the retain-history option to it if that sounds good to you.

AFAICT "create table ... from source` is not documented anywhere yet, so I have nothing to link to here. I'm open to suggestions!

Author: > As for the create table page -- that does have a with_options section, yes?

Not in the same way as: https://materialize.com/docs/sql/create-materialized-view/#with_options

The MV page has a general syntax for with options, then a table for the options that's separate from the main syntax. The table page has only a single option right now I guess... it's just defined alongside the other syntax and without the same level of detail. Happy to expand it out to an MV-style doc if that feels like a good change, though.

Reviewer: Oh ... yikes.  Yes, please.

```mzsql
CREATE TABLE events (created_at timestamptz, body jsonb)
WITH (
    PARTITION BY (created_at)
);
```

> When multiple columns are specified, rows are ordered lexicographically:

Author: Worked something like this in!

`PARTITION BY (created_date, created_time)` would sort first by the created date, then order rows with equal created date by time. Durable collections without a `PARTITION BY` option might be stored in any order.

> Note that this declaration has no impact on the order in which records are returned by queries. If you want to return results in a specific order, use an

Reviewer: Would just use a note box to draw more attention to this

```suggestion
{{< note >}}
The `PARTITION BY` option has no impact on the order in which records are returned by queries. If you want to return results in a specific order, use an `ORDER BY` clause on your select statement.
{{< /note >}}
```

We might also want to repeat the note up top because people will skip over it.

### Partitioning by timestamp

For timeseries or "event"-type collections, it's often useful to partition the data by timestamp.

1. First, create a table called `events`.
    ```mzsql
    -- Create a table of timestamped events. Note that the `event_ts` column is
    -- first in the column list and in the parition-by clause.
    CREATE TABLE events (
        event_ts timestamptz,
        content text
    ) WITH (
        PARTITION BY (event_ts)
    );
    ```

1. Insert a few records, one "older" record and one more recent.
    ```mzsql
    INSERT INTO events VALUES (now()::timestamp - '1 minute', 'hello');
    INSERT INTO events VALUES (now(), 'world');
    ```

1. If we run a select statement against the data sometime in the next couple of minutes, we return one row but not the other.
    ```mzsql
    SELECT * FROM events WHERE event_ts > mz_now() - '2 minutes';
    ```

1. An `EXPLAIN FILTER PUSHDOWN` statement shows that we're able to fetch a subset of the parts in our collection... only those parts that contain data with recent timestamps.
    ```mzsql
    EXPLAIN FILTER PUSHDOWN FOR
    SELECT * FROM events WHERE event_ts + '2 minutes' > mz_now();

>     ```

Author: Ideally we'd link to `EXPLAIN FILTER PUSHDOWN`, but that's not documented anywhere to my knowledge yet. That may make sense as a separate PR?

Reviewer: Actually, it is https://materialize.com/docs/sql/explain-filter-pushdown/

- If a materialized view has a temporal filter that preserves only the last two days of data, we'd like to filter out parts that contain data from a month ago.
- If a select statement filters for only data from a particular country, we'd like to avoid fetching parts that only contain data from other countries.

This optimization tends to be important for append-only timeseries datasets, but it can be useful for any dataset where most queries look at only a particular "range" of the data. If the query is a select statement, it can make that select statement more performant; if the query is an index or a materialized view, it can make it much faster to bootstrap.

Materialize will always try to filter out parts it doesn't need for a particular query, but that filtering is usually only effective when similar rows are stored together. If you want to make sure that filter pushdown reliably kicks in for your query, you should:

- Use a `PARTITION BY` clause on the relevant column to ensure that data with similar values for that column are stored close together.

> - Add a filter to your query that only returns true for a particular range of values in that column.

Author: I felt that the rewrite here was a bit unclear about the link between filters and the optimization, but I agree the first version was not quite right either. Tried a rewrite based on some of your ideas here in case any of that looks better?

- Only certain types of columns are supported. This includes:
    - all fixed-width integer types, including `smallint`, `integer`, and `bigint`;
    - date and time types, including `date`, `time`, `timestamp`, `timestamptz`, and `mz_timestamp`;
    - string types like `text` and `bytea`;
    - `boolean` and `uuid`;
    - `record` types where all fields types are supported.

We intend to relax some of these restrictions in the future.

## Examples

These examples create real objects. After you have tried the examples, make sure to drop these objects and spin down any resources you may have created.

### Partitioning by timestamp

For timeseries or "event"-type collections, it's often useful to partition the data by timestamp.

1. First, create a table called `events`.
    ```mzsql
    -- Create a table of timestamped events. Note that the `event_ts` column is
    -- first in the column list and in the parition-by clause.
    CREATE TABLE events (
        event_ts timestamptz,
        content text
    ) WITH (
        PARTITION BY (event_ts)
    );
    ```

1. Insert a few records, one "older" record and one more recent.
    ```mzsql

>     INSERT INTO events VALUES (now()::timestamp - '1 minute', 'hello');

Author: I've relaxed the timebound here!

    INSERT INTO events VALUES (now(), 'world');
    ```

1. If we run a select statement against the data sometime in the next couple of minutes, we return one row but not the other.
    ```mzsql
    SELECT * FROM events WHERE event_ts > mz_now() - '2 minutes';
    ```

1. An `EXPLAIN FILTER PUSHDOWN` statement shows that we're able to fetch a subset of the parts in our collection... only those parts that contain data with recent timestamps.
    ```mzsql
    EXPLAIN FILTER PUSHDOWN FOR
    SELECT * FROM events WHERE event_ts + '2 minutes' > mz_now();
    ```

If you wait a few minutes longer until there are no events that match the temporal filter, you'll notice that not only does the query return zero rows, but the explain shows that we fetched zero parts.

Note that the exact numbers you see here may very: parts can be much larger than a single row, and the actual level of filtering may vary for small datasets as data is compacted together internally. However, datasets of a few gigabytes or larger should reliably see benefits from this optimization.

### Partitioning by primary key

> Other datasets don't have a strong timeseries component, but they do have a clear notion of type or category. For example, let's suppose we manage a collection of music venues spread across the world, but regularly want to target queries to just those that exist in a single country.

Reviewer: The "For example" sentence, maybe just:

For example, suppose you have a collection of music venues spread across the world that you regularly query by a single country.

1. First, create a table called `venues`, partitioned by country.
    ```mzsql
    -- Create a table for our venue data.
    -- Once again, the partition column is listed first.
    CREATE TABLE venues (
        country_code text,
        id uuid,
    ) WITH (
        PARTITION BY (country_code)
    );
    ```

> 1. Insert a few records, one "older" record and one more recent.

Reviewer: Just a copy+paste error.

```suggestion
1. Insert a few records with different country codes.
```

    ```mzsql
    INSERT INTO venues VALUES ('US', uuid_generate_v5('venue', 'Rock World'));
    INSERT INTO venues VALUES ('CA', uuid_generate_v5('venue', 'Friendship Cove'));
    ```

> 1. We can filter down our list of venues to just those in a specific country.

Reviewer: ```suggestion
1. Query by a specific country.
```
Also, should we also show an example where if we had inserted data for  'US', 'CA', 'MX' and query for ('CA', 'MX')

Reviewer: I would expect that only the parts with CA and MX will return (they could possibly be the same parts if I guess we don't have a lot of data between CA and MX -- but not really something that concerns me here) ... but it's more that for some of our users who may not have deep SQL background, I don't want them to think they have to issue SELECT statement per country - but that partition can also help if you do `SELECT * FROM venues WHERE country_code in ('CA', 'MX')`.  That is, I don't want them to subconsciously limit the feature.

Author: Cool! That is all correct... I've tweaked the example.

    ```mzsql
    SELECT * FROM venues WHERE country_code = 'US';
    ```

> 1. An `EXPLAIN FILTER PUSHDOWN` statement shows that we're only fetching a subset of parts we've stored.

Reviewer: ```suggestion
1. To verify that Materialize fetched only the parts that contain data with the
   `US` country code, run `EXPLAIN FILTER PUSHDOWN`.
```

[...]
Source materialize.public.events
  [...]
  pushdown=((mz_now() <= timestamp_to_mz_timestamp((#0 + 00:05:00))))```
```

Some common functions, such as casting from a string to a timestamp, can prevent filter pushdown for a query. For similar functions that _do_ allow pushdown, see [the pushdown functions documentation](/sql/functions/pushdown/).

## Examples

These examples create real objects. After you have tried the examples, make sure to drop these objects and spin down any resources you may have created.

### Partitioning by timestamp

For timeseries or "event"-type collections, it's often useful to partition the data by timestamp.

1. First, create a table called `events`.
    ```mzsql
    -- Create a table of timestamped events. Note that the `event_ts` column is
    -- first in the column list and in the parition-by clause.
    CREATE TABLE events (
        event_ts timestamptz,
        content text
    ) WITH (
        PARTITION BY (event_ts)
    );
    ```

1. Insert a few records, one "older" record and one more recent.
    ```mzsql
    INSERT INTO events VALUES (now()::timestamp - '5 minutes', 'hello');
    INSERT INTO events VALUES (now(), 'world');
    ```

> 1. Run a select statement against the data sometime in the next few minutes. This should return only the more recent of the two rows.

Reviewer: We can remove the "sometime in the next few minutes" yes? It kind of makes it sound like people should wait ... as opposed to don't wait too long before running this ... but I think people should be able to tell.

We could explicitly state:
Run the following select statement to return events that occurred within the last two minutes.  This should return ...

Author: "Don't wait too long before running this" is definitely the point I want to emphasize. (Because if you wait too long the result changes.)

A rewrite like:

> Run the following select statement to return events that occurred within the last two minutes. This should return only the more recent of the two rows.

Would be incorrect: you need to say something about when the query is being run to say what the results will be.

Author: Yeah, they probably will! But I'd rather be precise now so I don't get a bug report later. 😅

I've switched to "within the next five minutes". Language like "immediately" would also be fine I think!

---
title: "Partitioning and filter pushdown"
description: "Declare how collections are stored."
aliases:
  - /guides/partition-by/
  - /sql/patterns/partition-by/
menu:
  main:
    parent: 'sql-patterns'
---

[//]: # "TODO link to the source table docs once that feature is documented."

{{< private-preview />}}

A few types of Materialize collections are durably written to storage: [materialized views](/sql/create-materialized-view/), [tables](/sql/create-table), and [sources](/sql/create-source).

> Internally, Materialize stores these durable collections in an [LSM-tree](https://en.wikipedia.org/wiki/Log-structured_merge-tree)-like structure. Each collection is made up of a set of

Reviewer 2: This is technically correct, but I wonder if we should simplify the explanation? If we just skipped the first 1 and 1/2 sencence and said something like
> Data is sorted and then partitioned up into individual **parts**, ...

And then later

>Using the `PARTITION BY` option, you can specify the internal ordering that Materialize will use to sort, partition, and store the data.

Author: The thing I worry about is questions like: "I know my query only matches a single record; if all the data is sorted, why does my query need to fetch more than one part?" Really hard to explain without the concept of "runs" here.

I'm definitely sensitive to this point - so far these are implementation details we haven't spoken much about publicly, and it makes sense to keep the description as light as possible. But I don't want to give the impression that everything is one sorted set of parts...

Reviewer 2: I see. Would it make sense to add a tiny example at the end that helps clarify the concepts and shows how filter pushdown can skip certain runs?

Author: I've dropped the first sentence since it's not strictly necessary, but kept a little language around runs. I'm inclined to leave it there for now; we can always follow up later as we learn more!

- This clause must list a prefix of the columns in the collection. For example:
  - if you're creating a table that partitions by a single column, that column must be the first column in the table's schema definition;
  - if you're creating a table that partitions by two columns, those columns must be the first two columns in the table's schema definition and listed in the same order.
- Only certain types of columns are supported. This includes:
    - all fixed-width integer types, including `smallint`, `integer`, and `bigint`;
    - date and time types, including `date`, `time`, `timestamp`, `timestamptz`, and `mz_timestamp`;
    - string types like `text` and `bytea`;
    - `boolean` and `uuid`;
    - `record` types where all fields types are supported.

We intend to relax some of these restrictions in the future.

## Filter pushdown

Suppose we're running a query against our `events` data with a [temporal filter](/transform-data/patterns/temporal-filters/).

```mzsql
SELECT * FROM events WHERE mz_now() <= event_ts + INTERVAL '5min';
```

This query returns only rows with similar values for `event_ts`: timestamps in the last five minutes.

> Since we declared that our `events` table is partitioned by `event_ts`, Materialize can determine that all the rows that pass this filter must also be _stored_ close together, and fetch only the exact parts it needs to satisfy the query.

Reviewer 2: I find this explanation a bit unsatisfying.

>and fetch only the exact parts it needs to satisfy the query.

It does not really explain how. That may be on purpose, which would be fine. But if we wanted to share some more details, we could add a sentence or two on how the metadata of parts can be used to prune parts without reading the data and how ordering makes this more effective.

Author: I think I can probably say a bit more here without getting _too_ deep in the weeds... I'll take another look.

Author: Took a stab at it!

Reviewer 2: Looks great, thanks!

    -- Create a table of timestamped events. Note that the `event_ts` column is
    -- first in the column list and in the parition-by clause.
    CREATE TABLE events (
        event_ts timestamptz,
        content text
    ) WITH (
        PARTITION BY (event_ts)
    );
    ```

1. Insert a few records, one "older" record and one more recent.
    ```mzsql
    INSERT INTO events VALUES (now()::timestamp - '5 minutes', 'hello');
    INSERT INTO events VALUES (now(), 'world');
    ```

1. Run a select statement against the data sometime in the next few minutes. This should return only the more recent of the two rows.
    ```mzsql
    SELECT * FROM events WHERE event_ts + '2 minutes' > mz_now();
    ```

1. To verify that Materialize fetched only the parts that contain data with the
   recent timestamps, run an `EXPLAIN FILTER PUSHDOWN` statement.
    ```mzsql
    EXPLAIN FILTER PUSHDOWN FOR
    SELECT * FROM events WHERE event_ts + '2 minutes' > mz_now();
    ```

If you wait a few minutes longer until there are no events that match the temporal filter, you'll notice that not only does the query return zero rows, but the explain shows that we fetched zero parts.

{{< note >}}

> The exact numbers you see here may very: parts can be much larger than a single row, and the actual level of filtering may vary for small datasets as data is compacted together internally. However, datasets of a few gigabytes or larger should reliably see benefits from this optimization.

Reviewer 2: typo: vary
